### DISCUSSION TITLE
CentOS 9 (Apple Silicon) PHP Images ready for Beta Testing!

### DISCUSSION THREAD

**Discussion by navarr** [wardenenv/warden#616] [Announcements]
Happy Wednesday, everyone!

The good news today is that the CentOS 9 PHP images are now ready for beta testing.  These images support both amd64 and arm64 operating systems - so it's imperative that both are well tested!

Please, _please_ help me test out these images and report any issues with them.  Ideally, I'd like to make these the primary images starting March 20th, assuming there are no major issues left unresolved.

## How to Test

Two steps must be taken to test out these images.  The first, is that you must be using the current in-development version of Warden.  To do so, you can either checkout the `develop` branch if you installed via git, or you can run `brew install wardenenv/warden/warden --HEAD` to have brew upgrade to the develop version.

After you have done this, you need to add to either your `~/.warden/.env` file (recommended) or a project's `.env` file: `WARDEN_NIGHTLY=1`.  If you have WARDEN_NIGHTLY set, Warden will instead use the `-indev` images.  You can verify this by running `warden env config` in an environment and checking the `image` underneath your php services.

## How to Report

Please report issues as normal through the [Warden New Issue Wizard](https://github.com/wardenenv/warden/issues/new/choose).

**Comment by wigman** [comment]
Awesome work @navarr. Just installed an instance and it works well. A lot faster on Mac M2 Max processor than the previous images. If I run into any issues later I'll report here.

**Reply by ihor-sviziev** [reply]
Also was testing it today - looks like everything works fine!

**Comment by Vinai** [comment]
After migrating as described above, I get the following error when running `warden env up`:
```
 % warden env up
Connecting traefik to mos-hyva-245p1_default network
Connecting tunnel to mos-hyva-245p1_default network
Connecting mailhog to mos-hyva-245p1_default network
ERROR: no such image: docker.io/wardenenv/mariadb:${MARIADB_VERSION:-10.4}: invalid reference format
```

It looks like bash variable expansion isn't being applied, and the literal string is used as the version.
Any ideas?

**Reply by Vinai** [reply]
@wigman had the idea to run the command using `bash` instead of the default shell `zsh`.
And: it worked! 🎉 

Maybe worth investigating how to support zsh, too, though.

**Reply by ihor-sviziev** [reply]
What `docker-compose -v` shows for you?

**Reply by Vinai** [reply]
Now the containers start, but requests aren't routed to the container correctly.  

After running `warden env up` in a project, when I enter the project URL, for example https://app.mos-hyva-245p1.test/ I get the same 404 that I get when I enter the domain for a non-active project (e.g. https://app.mos-hyva-241.test/)

DNS resolution through dnsmasq is working.

I can reach https://traefik.warden.test/, https://dnsmasq.warden.test/, https://portainer.warden.test/ and https://mailhog.warden.test/
I haven't regenerated warden's SSL certificates on the new machine yet, but I'm pretty sure that isn't the issue that is causing the routing to fail.

Any ideas how I could to debug this?

**Reply by Vinai** [reply]
@ihor-sviziev 
> What `docker-composer -v` shows for you?

```
$ docker-compose -v
docker-compose version 1.29.2, build 5becea4c
```

**Reply by ihor-sviziev** [reply]
I have the following version and everything works fine. 
```
Docker Compose version v2.15.1
``` 
So please check the "Use Docker Compose V2" in the Docker Desktop, and it will work
<img width="1382" alt="image" src="https://user-images.githubusercontent.com/1873745/225613673-419467b1-f3da-45e9-bd40-9bb110496fa4.png">

PS @navarr I think we should use `docker compose` (use compose v2 version) everywhere since the following notice is there:
https://docs.docker.com/compose/install/
> From the end of June 2023 Compose V1 won’t be supported anymore and will be removed from all Docker Desktop versions.
> Make sure you switch to [Compose V2](https://docs.docker.com/compose/compose-file/) with the docker compose CLI plugin or by activating the Use Docker Compose V2 setting in Docker Desktop. For more information, see the [Evolution of Compose](https://docs.docker.com/compose/compose-v2/)

**Reply by navarr** [reply]
I'm definitely using ZSH on my own computer and am not experiencing this issue.  I'd appreciate an issue filed with your `.env` file so I can take a closer look at it.  And just to verify you did brew install with the `--HEAD` attribute so it's using the `develop` branch, yes?

If you're not getting routed to your environments still after Ihor's idea, I'd double check whats happening in Traefik and ensure that your containers are, in fact, up after the up command is ran.

**Reply by navarr** [reply]
@ihor-sviziev It will be part of v1.  I did that in Den and I plan to port it over.

**Reply by ihor-sviziev** [reply]
Let's wait for answer from Vinai, 99% sure it's related to docker compose version

**Reply by Vinai** [reply]
Updating docker-compose worked, as in now `-v` shows 2.15.1

But it didn't help with the routing.

However, now, when I run `warden svc up` (or `warden env up`) I get a bunch of warnings:
```
 ⠏ portainer The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
 ⠏ dnsmasq The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
 ⠏ tunnel The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
 ⠏ traefik The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested   
 ```

Do I need to do additional steps to switch to arm64 containers?

**Reply by ihor-sviziev** [reply]
Try running `warden svc pull && warden svc up` AND make sure you've installed warden from develop branch using `brew install wardenenv/warden/warden --HEAD` (the `--HEAD` flag is very important)

**Reply by Vinai** [reply]
I definitely am using warden from the develop branch.

No difference after running `warden svc pull && warden svc up`.
However, I just noticed something else - the `varnish` container stops after a few seconds.
Maybe that is the reason the routing isn't working.

**Reply by navarr** [reply]
If you can, please drop your `.env` file and the log output of the Varnish container - which you should be able to grab from Docker Desktop if you're using it.  If not, _maybe_ `warden env log varnish` will have it even though it stopped.  That's definitely what's preventing the routing.

I'm curious how you're pulling amd64 instead of arm64 images.  I'm assuming you installed the Apple silicon docker, but I'm not sure how you would double check

**Reply by Vinai** [reply]
The project .env:
```
WARDEN_ENV_NAME=mos-hyva-245p1
WARDEN_ENV_TYPE=magento2
WARDEN_WEB_ROOT=/

TRAEFIK_DOMAIN=mos-hyva-245p1.test
TRAEFIK_SUBDOMAIN=app

WARDEN_DB=1
WARDEN_ELASTICSEARCH=1
WARDEN_VARNISH=1
WARDEN_RABBITMQ=1
WARDEN_REDIS=1

ELASTICSEARCH_VERSION=7.6
MARIADB_VERSION=10.4
NODE_VERSION=14
COMPOSER_VERSION=2
PHP_VERSION=8.1
PHP_XDEBUG_3=1
RABBITMQ_VERSION=3.8
REDIS_VERSION=5.0
VARNISH_VERSION=6.0

WARDEN_SYNC_IGNORE=

WARDEN_ALLURE=0
WARDEN_SELENIUM=0
WARDEN_SELENIUM_DEBUG=0
WARDEN_BLACKFIRE=0
WARDEN_SPLIT_SALES=0
WARDEN_SPLIT_CHECKOUT=0
WARDEN_TEST_DB=0
WARDEN_MAGEPACK=0

BLACKFIRE_CLIENT_ID=
BLACKFIRE_CLIENT_TOKEN=
BLACKFIRE_SERVER_ID=
BLACKFIRE_SERVER_TOKEN=
```

My `~/.warden.env`:
```
WARDEN_NIGHTLY=1
```

The varnish container log:
```
2023-03-16 13:31:39 Error:
2023-03-16 13:31:39 Message from VCC-compiler:
2023-03-16 13:31:39 Assert error in vju_subproc(), mgt/mgt_jail_unix.c line 212:
2023-03-16 13:31:39   Condition((initgroups(vju_user, vju_gid)) == 0) not true.
2023-03-16 13:31:39   errno = 1 (Operation not permitted)
2023-03-16 13:31:39 qemu: uncaught target signal 6 (Aborted) - core dumped
2023-03-16 13:31:39 Running VCC-compiler failed, signal 6
2023-03-16 13:31:39 VCL compilation failed
2023-03-16 13:38:10 Error:
2023-03-16 13:38:10 Message from VCC-compiler:
2023-03-16 13:38:10 Assert error in vju_subproc(), mgt/mgt_jail_unix.c line 212:
2023-03-16 13:38:10   Condition((initgroups(vju_user, vju_gid)) == 0) not true.
2023-03-16 13:38:10   errno = 1 (Operation not permitted)
2023-03-16 13:38:10 qemu: uncaught target signal 6 (Aborted) - core dumped
2023-03-16 13:38:10 Running VCC-compiler failed, signal 6
2023-03-16 13:38:10 VCL compilation failed
2023-03-16 13:38:31 Error:
2023-03-16 13:38:31 Message from VCC-compiler:
2023-03-16 13:38:31 Assert error in vju_subproc(), mgt/mgt_jail_unix.c line 212:
2023-03-16 13:38:31   Condition((initgroups(vju_user, vju_gid)) == 0) not true.
2023-03-16 13:38:31   errno = 1 (Operation not permitted)
2023-03-16 13:38:31 qemu: uncaught target signal 6 (Aborted) - core dumped
2023-03-16 13:38:31 Running VCC-compiler failed, signal 6
2023-03-16 13:38:31 VCL compilation failed
```

**Reply by navarr** [reply]
Quick point of order, most likely just a typo, but your WARDEN_NIGHTLY=1 is in `~/.warden/.env` not ~/.warden.env` right?

**Reply by Vinai** [reply]
> your WARDEN_NIGHTLY=1 is in `~/.warden/.env not ~/.warden.env` right?

Correct, sorry about that. It is in `~/.warden/.env`.

**Reply by Vinai** [reply]
Update: I removed *both* warden installations and then reinstalled the dev branch version.
Then I ran `warden svc pull && warden svc up`, folllowed by `warden env pull && warden env up`.

Now things are looking much different.

<img width="1104" alt="Bildschirm­foto 2023-03-17 um 06 44 49" src="https://user-images.githubusercontent.com/72463/225822592-d4b99c46-8657-472e-aeda-c3ace9940db4.png">

Only `elasticsearch-hq` is still an amd64 image.  

I can visit Magento in the browser.  
However, no products are visible.
Running `bin/magento ind:rein` gives the error

```
Catalog Search index process error during indexation process:
Could not ping search engine: No alive nodes found in your cluster
```

It seems something regarding elastic still needs tweaking...

**Reply by sprankhub** [reply]
elasticsearch-hq is old and not used any more AFAIK.

If you have issues with elastic, try removing the Elasticsearch volume with `docker volume rm {{VOLUME_NAME}}` and then restart the environment. The volume should then be recreated. Afterwards, reindex. Does that help?

**Reply by Vinai** [reply]
@sprankhub thanks, I'm currently unable to check - will do asap and report back 👍

**Reply by Vinai** [reply]
@sprankhub The elasticsearch-hq container is recreated with the amd image.
```
elasticsearch-hq The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested  0.0s
```
Indexation still fails with the same error as above: `Could not ping search engine: No alive nodes found in your cluster`

**Reply by ihor-sviziev** [reply]
About the elastic hq - it's optional now, you can turn it off by adding 0 to your .env file in project

Here is related commit https://github.com/wardenenv/warden/commit/f54aa952e97e5838144500f5ab8234dad2c8ad03

Edit: Oops, you've asked about rabbitmq, probably I didn't woke up yet. Unfortunately I'm not using it so it's turned off for me

**Reply by Vinai** [reply]
@ihor-sviziev thank  you 👍 
I asked about rabbit in a new project (that was in discord). 
Here in this thread it's a different project migrated from the old mpb.

**Reply by Vinai** [reply]
After adding `WARDEN_ELASTICHQ=0` to the project `.env` the container no longer is created.
No more amd64 container warnings 🎉 

However, I'm still not able to reindex:
```
Catalog Search index process error during indexation process:
Could not ping search engine: No alive nodes found in your cluster
```

**Reply by sprankhub** [reply]
Just to be sure... You did remove the Elasticsearch volume as described? What does [ES cluster health](https://stackoverflow.com/a/27377026/719023) tell you?

**Reply by Vinai** [reply]
@sprankhub Just to be sure I followed your instructions, I did `warden env stop && docker image rm mos-hyva-245p1-elasticsearch-1`.  
Then I ran `docker volume rm mos-hyva-245p1_esdata`  

After that I did
```
 % warden env up
Connecting traefik to mos-hyva-245p1_default network
Connecting tunnel to mos-hyva-245p1_default network
Connecting mailhog to mos-hyva-245p1_default network
[+] Running 9/9
 ⠿ Volume "mos-hyva-245p1_esdata"            Created
 ⠿ Container mos-hyva-245p1-elasticsearch-1  Started
 ⠿ Container mos-hyva-245p1-rabbitmq-1       Started
 ⠿ Container mos-hyva-245p1-db-1             Started
 ⠿ Container mos-hyva-245p1-redis-1          Started
 ⠿ Container mos-hyva-245p1-php-fpm-1        Started
 ⠿ Container mos-hyva-245p1-nginx-1          Started
 ⠿ Container mos-hyva-245p1-php-debug-1      Started
 ⠿ Container mos-hyva-245p1-varnish-1        Started
"docker container inspect" requires at least 1 argument.
See 'docker container inspect --help'.

Usage:  docker container inspect [OPTIONS] CONTAINER [CONTAINER...]

Display detailed information on one or more containers
```

I had not noticed that error earlier, I don't think it was there before.

**Reply by Vinai** [reply]
After another `warden env stop && warden env up` the error is gone, synchronization completes.
And finally:
```
$ bin/magento ind:rein
Design Config Grid index has been rebuilt successfully in 00:00:00
Customer Grid index has been rebuilt successfully in 00:00:00
Category Products index has been rebuilt successfully in 00:00:03
Product Categories index has been rebuilt successfully in 00:00:00
Catalog Rule Product index has been rebuilt successfully in 00:00:00
Product EAV index has been rebuilt successfully in 00:00:00
Inventory index has been rebuilt successfully in 00:00:00
Catalog Product Rule index has been rebuilt successfully in 00:00:00
Stock index has been rebuilt successfully in 00:00:00
Product Price index has been rebuilt successfully in 00:00:00
Catalog Search index has been rebuilt successfully in 00:00:18
```

My sample data is back ❤️ 

Thank you all very much for your help: 🙏

**Reply by Vinai** [reply]
So, to summarize, after migrating from an intel mac, I think these steps where necessary for me:

* [switch to docker-compose 2](https://github.com/wardenenv/warden/discussions/616#discussioncomment-5333083) in the Docker Desktop
* delete the old warden installation in /usr/local/Cellar
* install the new warden version as described at the [start of this discussion](https://github.com/wardenenv/warden/discussions/616#discussion-4942173)
* [add](https://github.com/wardenenv/warden/discussions/616#discussioncomment-5342419) `WARDEN_ELASTICHQ=0` to the project `.env`
* [remove the elasticsearch volume](https://github.com/wardenenv/warden/discussions/616#discussioncomment-5342144) using `docker volume rm {{VOLUME_NAME}}`
* run `warden svc pull && warden env pull` before running `warden svc up && warden env up`
* if there is an error `"docker container inspect" requires at least 1 argument.`, run `warden env stop && warden env up`
* run `warden shell` and then reindex with `bin/magento index:reindex`

**Comment by Vinai** [comment]
Setting up a new project on silicon, I'm encountering an issue with rabbitmq after the first `warden env up`

<img width="889" alt="Bildschirm­foto 2023-03-17 um 07 48 01" src="https://user-images.githubusercontent.com/72463/225867140-ccc9ccb9-7761-4aef-a716-23d24fd7f91b.png">

This is my `.env`
```
WARDEN_ENV_NAME=hyva-dev
WARDEN_ENV_TYPE=magento2
WARDEN_WEB_ROOT=/

TRAEFIK_DOMAIN=hyva-dev.test
TRAEFIK_SUBDOMAIN=app

WARDEN_DB=1
WARDEN_ELASTICSEARCH=0
WARDEN_OPENSEARCH=1
WARDEN_ELASTICHQ=0
WARDEN_VARNISH=1
WARDEN_RABBITMQ=1
WARDEN_REDIS=1

OPENSEARCH_VERSION=2.5
DB_DISTRIBUTION=mariadb
DB_DISTRIBUTION_VERSION=10.6
NODE_VERSION=18
COMPOSER_VERSION=2
PHP_VERSION=8.1
PHP_XDEBUG_3=1
RABBITMQ_VERSION=3.9
REDIS_VERSION=7.0
VARNISH_VERSION=7.1

WARDEN_SYNC_IGNORE=

WARDEN_ALLURE=0
WARDEN_SELENIUM=0
WARDEN_SELENIUM_DEBUG=0
WARDEN_BLACKFIRE=0
WARDEN_SPLIT_SALES=0
WARDEN_SPLIT_CHECKOUT=0
WARDEN_TEST_DB=0
WARDEN_MAGEPACK=0

BLACKFIRE_CLIENT_ID=
BLACKFIRE_CLIENT_TOKEN=
BLACKFIRE_SERVER_ID=
BLACKFIRE_SERVER_TOKEN=
```
(the only thing I changed was the node version)

And this is the rabbitmq container log:
```
2023-03-17 07:34:42 qemu: uncaught target signal 11 (Segmentation fault) - core dumped
2023-03-17 07:38:42 qemu: uncaught target signal 11 (Segmentation fault) - core dumped
2023-03-17 07:40:06 qemu: uncaught target signal 11 (Segmentation fault) - core dumped
```

**Reply by ihor-sviziev** [reply]
Very strange. Looks like you're using some old image for rabbitmq 3.9. Try running `warden env pull && warden env up`
I enabled rabbitmq for my project and it works fine w/o any emulation
<img width="1152" alt="image" src="https://user-images.githubusercontent.com/1873745/225892874-a528e8f2-6026-4c19-98f1-3e7d22a3d3b2.png">

**Reply by Vinai** [reply]
That helped. Thanks. Now proceeding with the setup.

**Reply by Vinai** [reply]
All containers are running the correct images.

I'm now following the [warden installation instructions](https://docs.warden.dev/environments/magento2.html) for a Magento 2.4.6 / PHP 8.2 dev environment.

When running the `bin/magento setup:install` command, the installation aborts with the error message
```
[Progress: 701 / 1406]
Installing search configuration...

In SearchConfig.php line 81:

  Could not validate a connection to Elasticsearch. No alive nodes found in your cluster
```

Are different arguments required when using opensearch maybe?

**Reply by Vinai** [reply]
Replaced the `elasticsearch` with the matching `opensearch` variants, but so far no success.  
Is `--search-engine=elasticsearch7` still correct?

**Reply by Vinai** [reply]
This seems to do the trick:
```
    --search-engine=opensearch \
    --opensearch-host=opensearch \
    --opensearch-port=9200 \
    --opensearch-index-prefix=magento2 \
    --opensearch-enable-auth=0 \
    --opensearch-timeout=15
```

**Reply by Vinai** [reply]
Now getting the error
```
[Progress: 1020 / 1406]
Module 'Magento_MsrpSampleData':
Enabling caches:
Current status:
layout: 1
block_html: 1
full_page: 1

In PatchApplier.php line 172:

  Unable to apply data patch Magento\MsrpSampleData\Setup\Patch\Data\InstallMsrpSampleData for module Magento_MsrpSampleData. Original exception message: Rolled back transaction has not been completed correctly.


In Mysql.php line 321:

  Rolled back transaction has not been completed correctly.
```

This is interesting. Installing Magento hasn't been like this since the early M2 days :)

**Reply by Vinai** [reply]
From `exception.log`
```
44 {main} {"exception":"[object] (OpenSearch\\Common\\Exceptions\\BadRequest400Exception(code: 429): {\"error\":{\"root_cause\":[{\"type\":\"cluster_block_exception\",\"reason\":\"index [magento2_product_1_v1] blocked by: [TOO_MANY_REQUESTS/12/
disk usage exceeded flood-stage watermark, index has read-only-allow-delete block];\"}],\"type\":\"cluster_block_exception\",\"reason\":\"index [magento2_product_1_v1] blocked by: [TOO_MANY_REQUESTS/12/disk usage exceeded flood-stage watermark, index has read-only-allow-delete block];\"},\"status\":429} at /var/www/html/vendor/opensearch-project/opensearch-php/src/OpenSearch/Connections/Connection.php:661)"} []
[2023-03-18T11:29:58.567924+00:00] main.CRITICAL: OpenSearch\Common\Exceptions\BadRequest400Exception: {"error":{"root_cause":[{"type":"cluster_block_exception","reason":"index [magento2_product_1_v1] blocked by: [TOO_MANY_REQUESTS/12/disk usage exceeded flood-stage watermark, index has read-only-allow-delete block];"}],"type":"cluster_block_exception","reason":"index [magento2_product_1_v1] blocked by: [TOO_MANY_REQUESTS/12/disk usage exceeded flood-stage watermark, index has read-only-allow-delete block];"},"status":429} in /var/www/html/vendor/opensearch-project/opensearch-php/src/OpenSearch/Connections/Connection.php:661
````

**Reply by Vinai** [reply]
Upping the docker virtual disk size in Docker Desktop to 300G fixed it.

**Reply by Vinai** [reply]
I'm now running into an issue where the `generated/code` directory is created with the user belonging to root

```
www-data@m2-hyva-246-php-fpm:/var/www/html$ ls -ld generated/ generated/code/
drwxr-xr-x 3 www-data www-data 4096 Mar 18 11:42 generated/
drwxrwxr-x 7 root     root     4096 Mar 18 11:42 generated/code/
www-data@m2-hyva-246-php-fpm:/var/www/html$
```

Trying to run `bin/magento setup:upgrade`  now triggers errors that generated classes can't be found (that is, they can not be written to the directory).

On further inspection I noticed the contents of `generated` are not synced to the host system. They are only present inside the php-fpm container.

**Reply by Vinai** [reply]
Restarting the container helped.
While doing though I noticed this message when running `warden svc up`:
```
% warden svc up
WARN[0000] Found orphan containers ([portainer]) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up.
```

Is this a problem that requires some action?

**Reply by ihor-sviziev** [reply]
This warning appears because you've disabled portainer container, but it wasn't removed, to do so you can run `warden svc up --remove-orphans`

**Reply by Vinai** [reply]
Thank you @ihor-sviziev 👍 
Currently I think I now have a fully functional migrated project as well as a new project. I'm very happy :)

**Comment by navarr** [comment]
Beta Testing is now closed.

Please see https://github.com/orgs/wardenenv/discussions/620
